node/src/components/consensus/highway_core/highway.rs

```diff
@@ -506,6 +511,25 @@ impl<C: Context> Highway<C> {
         }
     }
 
+    /// Adds evidence to the protocol state.
+    /// Gossip the evidence if it's the first equivocation from the creator.
+    fn add_evidence(
+        &mut self,
+        evidence: Evidence<C>,
+        rng: &mut dyn CryptoRngCore,
+    ) -> Vec<Effect<C>> {
+        let was_honest = !self.state.is_faulty(evidence.perpetrator());
+        self.state.add_evidence(evidence.clone());
+        let mut effects = self.on_new_evidence(&evidence, rng);
```

Comment:
That should also be inside the `if was_honest` block, shouldn't it?
And maybe `State::add_evidence` could return a `bool`, just like `HashSet::insert` does?

And most importantly, we need to call `on_new_evidence` also in `add_valid_vote` below, in the `if was_honest` block.

Reply:
> That should also be inside the `if was_honest` block, shouldn't it?

That's not necessary (b/c it just won't have any effect) but you're right, it will be better there.

> And maybe `State::add_evidence` could return a bool, just like `HashSet::insert` does?

None of the `State::add_*` methods do that and I don't think there's a need for that?

> And most importantly, we need to call `on_new_evidence` also in `add_valid_vote` below, in the `if was_honest` block.

I will look into that.

Reply:
> I don't think there's a need for that?

No, it would just allow you to get rid of the `was_honest` variable:
```rust
if self.state.add_evidence(evidence.clone()) {
```

Reply:
> And most importantly, we need to call `on_new_evidence` also in `add_valid_vote` below, in the `if was_honest` block.

That's taken care of in [`active_validator::on_new_vote`](https://github.com/CasperLabs/casper-node/pull/481/files#diff-dddbe85faff74747d137a8660059e2b9cc7c8057074a0c3cc689a30fbb0dda16R154-R159).

Reply:
I don't think it is: That will only apply to future votes that are added.
In `on_new_evidence`, we endorse _existing_ votes, because we just realized someone is faulty. There are two ways for us to realize someone is faulty:
* We receive evidence from another node — that's where you already call `on_new_evidence`.
* [We detect an equivocation.](https://github.com/CasperLabs/casper-node/blob/31251c10763da1edffbbfca4c9fc216ae17151d7/node/src/components/consensus/highway_core/state.rs#L588) That happens inside `add_valid_vote`, but inside of `State` without being reported. That's why you added [this check](https://github.com/CasperLabs/casper-node/blob/31251c10763da1edffbbfca4c9fc216ae17151d7/node/src/components/consensus/highway_core/highway.rs#L527) so we can gossip the new evidence anyway. In addition to that, we need to call `on_new_evidence`, I think, to create endorsements for existing votes?

Reply:
Ah, you're right :facepalm: In `active_validator::on_new_vote` I would only endorse an incoming vote (if needed) but not other votes if it happens to create an equivocation

Reply:
Should be done now.